# **CHECK POINT**
## **Disciplinas: Machine Learning & Modelling and Statistical Computing with R & Python**
- RM 574218 - GONZALO DURAN
- RM 575401 - MURILO WENGRZYNEK
- RM 575397 - ANDRÉ CHACON
- RM 575773 - MATHEO GRITTI VASCONCELLOS


# Parte 1 — Statistical Computing with R and Python

## 1. Coleta e carregamento dos dados

Para este trabalho vamos utilizar o **Iris Dataset**, que será a nossa base de dados para as duas partes do projeto.

O Iris é um dataset que possui informações de **150 flores**, divididas em três espécies:

- Iris Setosa
- Iris Versicolor
- Iris Virginica

Para cada flor temos quatro medidas, todas em centímetros:

- comprimento da sépala;
- largura da sépala;
- comprimento da pétala;
- largura da pétala.

Além dessas informações, cada registro possui um `target`, que identifica a espécie da flor.

O dataset será carregado diretamente da biblioteca Scikit-learn utilizando o `load_iris()`.

In [1]:
# Importação das bibliotecas que serão utilizadas ao longo do projeto

# Biblioteca para manipulação e análise de dados
import pandas as pd

# Função utilizada posteriormente para separar os dados em treino e teste
from sklearn.model_selection import train_test_split

# Algoritmo K-Nearest Neighbors (KNN), que será utilizado na Parte 2
from sklearn.neighbors import KNeighborsClassifier


### 1.1 Carregando o Iris

Primeiro carregamos o dataset em uma variável chamada `iris`.

Assim podemos acessar os dados, o target e as outras informações que fazem parte do dataset.

In [2]:
# Importação do dataset Iris
from sklearn.datasets import load_iris

# Carregamento do conjunto de dados
iris = load_iris()

# Visualização inicial do objeto
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

### 1.2 Conhecendo a estrutura dos dados

Após o carregamento, vamos conhecer um pouco melhor a estrutura do dataset, verificando as variáveis disponíveis, as espécies e a quantidade de registros.

In [3]:
# Visualização dos nomes das variáveis
print("\nVariáveis:")
print(iris.feature_names)

# Visualização das espécies presentes no dataset
print("\nEspécies:")
print(iris.target_names)

# Dimensão do conjunto de dados
print("\nDimensão dos dados:")
print(iris.data.shape)


Variáveis:
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

Espécies:
['setosa' 'versicolor' 'virginica']

Dimensão dos dados:
(150, 4)


A partir dessa primeira exploração, verificamos que o dataset possui **150 registros de flores** e **4 características** para cada registro. Também identificamos três espécies: **Setosa, Versicolor e Virginica**.

Nas próximas etapas vamos organizar esses dados em uma tabela e continuar a análise.

### 1.3 Organização dos dados

Para facilitar a exploração e a validação dos dados, vamos organizar o Iris Dataset em um DataFrame utilizando o Pandas.

As quatro características das flores serão utilizadas como colunas e também vamos adicionar a coluna `target`, que identifica a espécie de cada registro.

In [4]:
# Criação do DataFrame com os dados do Iris
df = pd.DataFrame(iris.data, columns=iris.feature_names)

# Adição da coluna target
df['target'] = iris.target

# Visualização das primeiras linhas
df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


### 1.4 Verificação de valores ausentes

Uma das etapas da validação dos dados é verificar se existem valores ausentes.

Valores ausentes podem prejudicar a análise dos dados e também o treinamento do modelo de Machine Learning.

In [5]:
# Verificação de valores ausentes em cada coluna
df.isnull().sum()

,0
sepal length (cm),0
sepal width (cm),0
petal length (cm),0
petal width (cm),0
target,0


O resultado mostra que não existem valores ausentes no dataset.

Como todas as colunas apresentaram valor 0, não foi necessário realizar nenhum tratamento para valores ausentes.

### 1.5 Verificação de duplicidades

Também vamos verificar se existem registros duplicados no dataset.

Para isso, podemos comparar as linhas do DataFrame e identificar quantas aparecem mais de uma vez.

In [6]:
# Verificação da quantidade de registros duplicados
df.duplicated().sum()

np.int64(1)

In [7]:
# Visualização do registro duplicado
df[df.duplicated()]

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
142,5.8,2.7,5.1,1.9,2


In [8]:
# Remoção do registro duplicado
df = df.drop_duplicates()

# Verificação após o tratamento
df.duplicated().sum()

np.int64(0)

Foi identificado **1 registro duplicado** no dataset.

Como tratamento, utilizamos o `drop_duplicates()` para remover a duplicidade. Após a remoção, realizamos uma nova verificação e não foram encontrados outros registros duplicados.

### 1.6 Verificação de inconsistências

Por último, vamos verificar possíveis inconsistências nos dados.

Como as quatro características das flores representam medidas em centímetros, valores iguais ou menores que zero poderiam indicar uma inconsistência.

Também vamos verificar os valores existentes na coluna `target`, que deve conter somente as três classes do Iris: 0, 1 e 2.

In [9]:
# Verificação de valores iguais ou menores que zero
print((df[iris.feature_names] <= 0).sum())

# Verificação dos valores existentes no target
print("\nValores do target:")
print(df['target'].value_counts())

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
dtype: int64

Valores do target:
target
0    50
1    50
2    49
Name: count, dtype: int64


Não foram encontrados valores iguais ou menores que zero nas características das flores.

Também verificamos que a coluna `target` possui somente as três classes esperadas: 0, 1 e 2.

Portanto, não foram identificadas inconsistências que precisassem de tratamento.

# Parte 2 — Machine Learning & Modelling

## 2. Preparação dos dados

Após a exploração, validação e tratamento realizados na Parte 1, o dataset está pronto para ser utilizado no modelo de Machine Learning.

Primeiro, vamos separar as características das flores do resultado que queremos prever.

Em `X` ficam as quatro características das flores e em `y` fica o `target`, que representa a espécie de cada registro.

In [10]:
# X recebe todas as características das flores
X = df.drop('target', axis=1)

# y recebe somente o target
y = df['target']

In [11]:
# Visualização das primeiras linhas de X
X.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


In [12]:
# Visualização dos primeiros valores de y
y.head()

,target
0,0
1,0
2,0
3,0
4,0


## 2.1 Separação dos dados de treino e teste

Depois de preparar `X` e `y`, vamos separar os dados em dois conjuntos.

Vamos utilizar **70% dos dados para treino** e **30% para teste**.

O `random_state=42` permite manter a mesma divisão dos dados cada vez que executamos o código.

In [13]:
# Separação dos dados de treino e teste
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

In [14]:
# Verificação do tamanho dos conjuntos
print("X treino:", X_train.shape)
print("X teste:", X_test.shape)
print("y treino:", y_train.shape)
print("y teste:", y_test.shape)

X treino: (104, 4)
X teste: (45, 4)
y treino: (104,)
y teste: (45,)


Após a divisão, temos **104 registros para treino** e **45 registros para teste**.

Os dados de treino serão utilizados para treinar o modelo e os dados de teste serão utilizados para realizar as previsões.

## 2.2 Treinamento do modelo KNN

Com os dados separados, podemos começar o treinamento do modelo **KNN (K-Nearest Neighbors)**.

Neste trabalho vamos testar diferentes valores de `k`: **1, 3, 5, 7 e 9**.

Para cada valor, vamos criar um modelo, realizar o treinamento com `fit()` e depois fazer a predição dos dados de teste com `predict()`.

In [21]:
# Modelo KNN com k = 1
modelo_k1 = KNeighborsClassifier(n_neighbors=1)

# Treinamento
modelo_k1.fit(X_train, y_train)

# Predição
ypred_k1 = modelo_k1.predict(X_test)

In [22]:
# Modelo KNN com k = 3
modelo_k3 = KNeighborsClassifier(n_neighbors=3)

# Treinamento
modelo_k3.fit(X_train, y_train)

# Predição
ypred_k3 = modelo_k3.predict(X_test)

In [23]:
# Modelo KNN com k = 5
modelo_k5 = KNeighborsClassifier(n_neighbors=5)

# Treinamento
modelo_k5.fit(X_train, y_train)

# Predição
ypred_k5 = modelo_k5.predict(X_test)

In [24]:
# Modelo KNN com k = 7
modelo_k7 = KNeighborsClassifier(n_neighbors=7)

# Treinamento
modelo_k7.fit(X_train, y_train)

# Predição
ypred_k7 = modelo_k7.predict(X_test)

In [25]:
# Modelo KNN com k = 9
modelo_k9 = KNeighborsClassifier(n_neighbors=9)

# Treinamento
modelo_k9.fit(X_train, y_train)

# Predição
ypred_k9 = modelo_k9.predict(X_test)

### Visualização das predições

Depois de treinar os modelos, podemos comparar algumas das predições realizadas com os valores reais do conjunto de teste.

Abaixo utilizamos o modelo com `k = 5` para visualizar essa comparação.

In [26]:
# Comparação entre as predições e os valores reais
print('pred   :', ypred_k5[:20])
print('y_test :', y_test[:20].values)

pred   : [1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2]
y_test : [1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2]


## 2.3 Cálculo da acurácia

Depois de realizar as previsões, vamos calcular a acurácia de cada modelo.

A acurácia representa a proporção de previsões corretas em relação ao total de registros utilizados no teste.

In [27]:
# Cálculo da acurácia para cada valor de k
acuracia_k1 = (ypred_k1 == y_test).mean()
acuracia_k3 = (ypred_k3 == y_test).mean()
acuracia_k5 = (ypred_k5 == y_test).mean()
acuracia_k7 = (ypred_k7 == y_test).mean()
acuracia_k9 = (ypred_k9 == y_test).mean()

# Visualização dos resultados
print("Acurácia k = 1:", acuracia_k1)
print("Acurácia k = 3:", acuracia_k3)
print("Acurácia k = 5:", acuracia_k5)
print("Acurácia k = 7:", acuracia_k7)
print("Acurácia k = 9:", acuracia_k9)

Acurácia k = 1: 1.0
Acurácia k = 3: 1.0
Acurácia k = 5: 1.0
Acurácia k = 7: 1.0
Acurácia k = 9: 1.0


## 2.4 Tabela comparativa

Por último, vamos organizar as acurácias em uma tabela para facilitar a comparação dos resultados obtidos com cada valor de `k`.

In [28]:
# Criação da tabela comparativa
resultados = pd.DataFrame({
    'k': [1, 3, 5, 7, 9],
    'Acurácia': [
        acuracia_k1,
        acuracia_k3,
        acuracia_k5,
        acuracia_k7,
        acuracia_k9
    ]
})
resultados

,k,Acurácia
0,1,1.0
1,3,1.0
2,5,1.0
3,7,1.0
4,9,1.0


## 2.5 Análise dos resultados

A tabela permite comparar a acurácia obtida pelo modelo KNN para cada valor de `k` testado.

A partir dos resultados, podemos observar se a alteração do número de vizinhos gerou diferenças nas previsões realizadas pelo modelo.